tests/src/test/scala/ha/ShootComponentsTests.scala

```diff
@@ -0,0 +1,151 @@
+
+  def restartComponent(host: String, component: String) = {
+    def file(path: String) = Try(new File(path)).filter(_.exists).map(_.getAbsolutePath).toOption
+    val docker = (file("/usr/bin/docker") orElse file("/usr/local/bin/docker")).getOrElse("docker")
+
+    val cmd = Seq(docker, "--host", host, "restart", component)
+    println(s"Running command: ${cmd.mkString(" ")}")
+
+    TestUtils.runCmd(0, new File("."), cmd: _*)
+  }
+
+  def ping(host: String, port: Int) = {
+    val response = Try { Http().singleRequest(HttpRequest(uri = s"http://$host:$port/ping")).futureValue }.toOption
+
+    response.map { res =>
+      (res.status, Unmarshal(res).to[String].futureValue)
+    }
+  }
+
+  def isControllerAlive(instance: Int): Boolean = {
+    require(instance >= 0 && instance < 2, "Controller instance not known.")
+
+    val host = WhiskProperties.getProperty("controller.hosts").split(",")(instance)
+    val port = WhiskProperties.getControllerBasePort + instance
+
+    val res = ping(host, port)
+    res.isDefined && res.get._1 == StatusCodes.OK && res.get._2 == "pong"
+  }
+
+  def doRequests(amount: Int, actionName: String): Seq[(Int, Int)] = {
+    (0 until amount).map { i =>
+      val start = Instant.now
+
+      // Do POSTs and GETs
+      val invokeExit = Future { wsk.action.invoke(actionName, expectedExitCode = TestUtils.DONTCARE_EXIT).exitCode }
+      val getExit = Future { wsk.action.get(actionName, expectedExitCode = TestUtils.DONTCARE_EXIT).exitCode }
+
+      println(s"Done rerquests with responses: invoke: ${invokeExit.futureValue} and get: ${getExit.futureValue}")
+
+      // Do at most one action invocation per second to avoid getting 429s. (60 req/min - limit)
+      val wait = 1000 - (Instant.now.toEpochMilli - start.toEpochMilli)
+      Thread.sleep(if (wait < 0) 0L else if (wait > 1000) 1000L else wait)
+      (invokeExit.futureValue, getExit.futureValue)
+    }
+  }
+
+  behavior of "Controllers hot standby"
+
+  it should "use controller1 if controller0 goes down" in withAssetCleaner(wskprops) { (wp, assetHelper) =>
+    if (WhiskProperties.getProperty(WhiskConfig.controllerInstances).toInt >= 2) {
+      val actionName = "shootcontroller"
+
+      assetHelper.withCleaner(wsk.action, actionName) { (action, _) =>
+        action.create(actionName, defaultAction)
+      }
+
+      // Produce some load on the system for 90 seconds. Kill the controller after 4 requests
+      val requestsBeforeRestart = doRequests(4, actionName)
+
+      // Kill the controller
+      val dockerHost = WhiskProperties.getBaseControllerHost() + ":" + WhiskProperties.getProperty(
+        WhiskConfig.dockerPort)
+      restartComponent(dockerHost, "controller0")
+      // Wait until down
+      retry({
+        isControllerAlive(0) shouldBe false
+      }, 100, Some(100.milliseconds))
+      // Check that second controller is still up
+      isControllerAlive(1) shouldBe true
+
+      val requestsAfterRestart = doRequests(96, actionName)
```

Comment:
Would it make sense to formulate this as `100 - 4` (define a `val totalRequests = 100` above?)